cli/src/commands/operation/diff.rs

```diff
@@ -0,0 +1,527 @@
+        from_op =
+            op_walk::resolve_op_for_load(repo_loader, args.from.as_ref().unwrap_or(head_op_str))?;
+        to_op = op_walk::resolve_op_for_load(repo_loader, args.to.as_ref().unwrap_or(head_op_str))?;
+    } else {
+        to_op = op_walk::resolve_op_for_load(
+            repo_loader,
+            args.operation.as_ref().unwrap_or(head_op_str),
+        )?;
+        let to_op_parents: Vec<_> = to_op.parents().try_collect()?;
+        if to_op_parents.is_empty() {
+            return Err(user_error("Cannot diff operation with no parents"));
+        }
+        from_op = repo_loader.merge_operations(command.settings(), to_op_parents, None)?;
+    }
+    let with_content_format = LogContentFormat::new(ui, command.settings())?;
+
+    let from_repo = repo_loader.load_at(&from_op)?;
+    let to_repo = repo_loader.load_at(&to_op)?;
+
+    // Create a new transaction starting from `to_repo`.
+    let mut workspace_command =
+        command.for_loaded_repo(ui, command.load_workspace()?, to_repo.clone())?;
+    let mut tx = workspace_command.start_transaction();
+    // Merge index from `from_repo` to `to_repo`, so commits in `from_repo` are
+    // accessible.
+    tx.mut_repo().merge_index(&from_repo);
+
+    ui.request_pager();
+    ui.stdout_formatter().with_label("op_log", |formatter| {
+        write!(formatter, "From operation ")?;
+        write!(
+            formatter.labeled("id"),
+            "{}",
+            short_operation_hash(from_op.id()),
+        )?;
+        write!(formatter, ": ")?;
+        write!(
+            formatter.labeled("description"),
+            "{}",
+            if from_op.id() == from_op.op_store().root_operation_id() {
+                "root()"
+            } else {
+                &from_op.metadata().description
+            }
+        )?;
+        writeln!(formatter)?;
+        write!(formatter, "  To operation ")?;
+        write!(
+            formatter.labeled("id"),
+            "{}",
+            short_operation_hash(to_op.id()),
+        )?;
+        write!(formatter, ": ")?;
+        write!(
+            formatter.labeled("description"),
+            "{}",
+            if to_op.id() == to_op.op_store().root_operation_id() {
+                "root()"
+            } else {
+                &to_op.metadata().description
+            }
+        )?;
+        writeln!(formatter)?;
+        Ok(())
+    })?;
+
+    show_op_diff(
+        ui,
+        command,
+        tx,
+        &from_repo,
+        &to_repo,
+        !args.no_graph,
+        &with_content_format,
+        &args.diff_format,
+        args.patch,
+    )
+}
+
+// Computes and shows the differences between two operations, using the given
+// `ReadonlyRepo`s for the operations.
+// `tx` should contain a `MutableRepo` with the indices of both repos merged
+// into it.
+#[allow(clippy::too_many_arguments)]
+pub fn show_op_diff(
+    ui: &mut Ui,
+    command: &CommandHelper,
+    mut tx: WorkspaceCommandTransaction,
```

Comment:
nit: perhaps, `tx` can be replaced with `repo: &dyn Repo` and `commit_summary_template`. It doesn't make sense that `cmd_op_show()` creates `tx`, and the summary template can be reused.

We can at least make it `&WorkspaceCommandTransaction` to clarify that this function does never mutate repo.

Reply:
This patch might help to build a template instance:
https://github.com/martinvonz/jj/compare/main...yuja:jj:push-nrkykowtyoor